[[ tsNet ]] Implement changes required to support tsNet libURL driver

montegoulding opened the pull request:
This patch provides support for the inclusion of
tsNet into the livecode package. All additional
externals (mergExt & tsNet) will now be found
in the same `Tools/Ext` directory.

Support has been implemented for a directory within
`Tools/Ext` to contain livecode files that may insert
themselves into the IDE message path and standalone
build process. This was required for tsNet's script
based libURL driver. These files will have
`revLoadLibrary` and `revUnloadLibrary` sent to
them so they may manage their own lifecycle as
back script, front script, library, behavior or
just loaded stack.

In the course of development a bug was identified
and resolved where android externals were potentially
being filtered from the included files. The
`revSBLibrary` was elevated to an always loaded
back script in the IDE due to the many scripts
depending on it.

There are dependencies between this patch and an
associated patch to the livecode-ide repo.

Commit:
[[ tsNet ]] Implement changes required to support tsNet libURL driver

This patch provides support for the inclusion of
tsNet into the livecode package. All additional
externals (mergExt & tsNet) will now be found
in the same `Tools/Ext` directory.

Support has been implemented for a directory within
`Tools/Ext` to contain livecode files that may insert
themselves into the IDE message path and standalone
build process. This was required for tsNet's script
based libURL driver. These files will have
`revLoadLibrary` and `revUnloadLibrary` sent to
them so they may manage their own lifecycle as
back script, front script, library, behavior or
just loaded stack.

In the course of development a bug was identified
and resolved where android externals were potentially
being filtered from the included files. The
`revSBLibrary` was elevated to an always loaded
back script in the IDE due to the many scripts
depending on it.

There are dependencies between this patch and an
associated patch to the livecode-ide repo.

techstrategies commented on builder/builder_utilities.livecodescript:
```diff
@@ -547,6 +552,35 @@ end macPList
+   if there is not a file pZipfile then
+      builderLog "message", "Downloading zip file:" && pZipfile
+      get shell(merge("curl -o '[[pPath]]' '[[pUrl]]'"))
+      if the result is not zero then
```
Should pPath be pZipfile here?

montegoulding replied:
Yes it should! Thanks for spotting that... Looking forward to implementing some basic refactoring in the script editor.

techstrategies commented on builder/builder_utilities.livecodescript:
```diff
@@ -555,42 +589,54 @@ private function builderMergExtDownloadedFilePath pMergExtVersion, pEdition
+   
+   if pEdition is "community" then
+      -- Ensure that the TSNet bundle is downloaded and unpacked
+      builderTSNetUnpack builderTSNetVersion()
```
Is this meant to be:
`if pEdition is not "community" then`

montegoulding replied:
@techstrategies Yes it is! This PR is not quite ready for review which is why it is labeled WIP but feel free to spot stuff like this definitely!

techstrategies replied:
@montegoulding That's cool, I figured you were still working on it.  I was just having a quick look over it and noticed those two things, so thought I'd highlight them for you!

montegoulding replied:
Thanks

montegoulding commented on ide-support/revsblibrary.livecodescript:
```diff
@@ -4,18 +4,16 @@
 #?author
 # Marcus van Houdt
 
```
😫 whitespace diffs that I didn't notice. I'll see what I can do to clean this up.

Commit:
Merge commit 'b558abb21128b34b977d08536f892ec49e68216a' into tsnet

# Conflicts:
#	builder/builder_utilities.livecodescript

techstrategies commented on ide-support/revsblibrary.livecodescript:
```diff
@@ -1838,16 +1888,52 @@ command revSBUpdateDeployParams pStack, pStandaloneSettings, @xDeployParams
+                  if tLoadedLibraries[tStackName] then
+                     appendToStringList the effective filename of stack tLibraryStack, tAuxiliaryStackfiles
+                     appendToStringList "start using stack" && quote & tStackName & quote, tStartupScript
+                  end if
```
@montegoulding Like my comment to your IDE PR, should this be calling revLoadLibrary rather than start using stack?

montegoulding replied:
Yep ;-)

techstrategies replied:
ok... just wanted to make sure I was following where you were going with this :-)

Commit:
[[ tsNet ]] Fix minor issues with tsNet integration

Commit:
[[ tsNet ]] Add URL tests for tsNet

montegoulding commented on builder/builder_utilities.livecodescript:
```diff
@@ -555,42 +589,54 @@ private function builderMergExtDownloadedFilePath pMergExtVersion, pEdition
+////////////////////////////////////////////////////////////////////////////////
+
+function builderTSNetUrl pTSNetVersion
+   return "https://downloads.livecode.com/tsNet/tsNet_" & pTSNetVersion & ".zip"
```
@techstrategies you have one zip with an underscore and one with a dash. The more recent one uses a dash. Which would you like to use?

techstrategies replied:
Yes, those need to be cleaned up, I will re-push when I finish some other changes.  I think it would be best to keep to the same format that is being used for mergExt.

montegoulding replied:
OK, I won't change the scripts or tests then. When you post an update the procedure will be to edit the version constant and post a PR for that after upload. Travis then runs the test to ensure the file is available. It also checks the version is a date but that really was only for mergExt because it is a whole bundle of individually versioned externals. If you want to use a version number that might be better and we can just drop the date check.

montegoulding replied:
I have renamed the one with the dash so they match and should pass the current tests

montegoulding commented:
Leaving as WIP while I sort out branches with the correct submodule pointers so I can test this. Also not sure if @techstrategies still wants to add linux builds before this is integrated.

Commit:
[[ tsNet ]] Revert changes to resave script only as binary handler

Commit:
[[ tsNet ]] Fix tsNet unpack folder as archive contains folder

Commit:
[[ tsNet ]] Only clean unpack folder once while unpacking downloads

Commit:
[[ tsNet ]] Remove requirement for tsNet pointer to be a date

Commit:
[[ tsNet ]] Various fixed for tsNet integration

- strict compilation of libURL
- ensure revLoadLibrary only handled for the correct target

Commit:
Merge commit 'a1a845871c36434f31857de9f137ec6b9eb74b33' into tsnet

# Conflicts:
#	builder/builder_utilities.livecodescript
#	ide-support/revliburl.livecodescript
#	ide-support/revsblibrary.livecodescript

Commit:
[[ tsNet ]] Support separate downloads per edition

Commit:
[[ tsNet ]] Support resources directory on android

Commit:
[[ tsNet ]] Update tsNet url test to support multiple editions

Commit:
[[ tsNet ]] Fix issues in resolving copy files

Commit:
Remove incorrect checks for `!` prefix

montegoulding commented:
For convenience in review I have merged all dependant PRs into a test branch which will be available from staging [here](https://downloads.livecode.com/livecode/staging/8.1.0.14001%20%28DP%202%29/gc09d5b7/)

As a reminder PR dependencies are:
- #4238 [[ Bug ]] Support multiple architectures for Ext externals
- livecode/livecode-ide#1244 [[ ExternalDependencies ]] Support LCS based dependencies for externals
-  #4240 [[ Bug 17958 ]] Ignore lcidl 'tail' attribute on desktop
-  #4239 [[ Bug 17957 ]] Fetch 'it' correctly in V1 external interface.

Related but not dependant:
- livecode/livecode-ide#1268 Check the target is me in revLoadLibrary and revUnloadLibrary

Commit:
Use platform instead of target on mobile when updating deploy params

livecodeali commented:
@livecode-vulcan review ok 2a99a70

livecode-vulcan commented:
💙 review by @livecodeali ok 2a99a709411c21e48654cb234dc06a5c61038d8d

livecode-vulcan commented:
😎 test success 2a99a709411c21e48654cb234dc06a5c61038d8d
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2763)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2772)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2758)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2774)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2780)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2784)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2802)
